File: 2. Add Two Numbers.py
# Definition for singly-linked list.
# Problem Link: https://leetcode.com/problems/add-two-numbers/
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # Time Complexity = O(m+n) memory complexity = O(1)
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = ListNode(0)
        head = res
        carry = 0
        while True:
            tot = carry
            carry = 0
            if l1:
                tot += l1.val
                l1 = l1.next
            if l2:
                tot += l2.val
                l2 = l2.next

            if tot >= 10:
                tot -= 10
                carry = 1

            res.val = tot
            if not l1 and not l2 and carry == 0:
                return head
            res.next = ListNode(0)
            res = res.next

File: C#/2. Add Two Numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null){
            return l2;
        }
        else if(l2 == null){
            return l1;
        }
        
        int len1 = 0, len2 = 0, carry = 0;
        ListNode temp1 = l1, temp2 = l2, large, small, toReturn;
        
        while(temp1 != null){
            len1++;
            temp1 = temp1.next;
        }
        
        while(temp2 != null){
            len2++;
            temp2 = temp2.next;
        }
        
        if(len1 >= len2){
            large = l1;
            small = l2;
        }
        else {
            large = l2;
            small = l1;
        }
        
        toReturn = large;
        
        while(true){
            int sum = carry;
            carry = 0;
            
            if (large != null){
                sum += large.val;
            }
            
            if (small != null){
                sum += small.val;
                small = small.next;
            }
            
            if (sum >= 10){
                sum -= 10;
                carry = 1;
            }
            
            large.val = sum;
            if (large.next == null && carry == 1){
                large.next = new ListNode(carry);
                break;
            }
            
            if (large != null){
                large = large.next;
            }
            
        }
        
        return toReturn;
    }
}